### PR TITLE
JQ filtering in client

### PR DESCRIPTION
**Pull request by gergas3**
Add ability to specify a `jq` filter that is applied to the response body. Allows extracting only parts of the response, which is useful when the response is verbose and the server is only interested in specific part of the response. An example use case is listing Kubernetes resources where the metadata of each item in the response is very verbose. (See test case for a concrete example.) Uses the [node-jq](https://www.npmjs.com/package/node-jq) library.
The jq tranform expression is passed in the `braekhus-jq-response-transform` header - not prefixed `x-` as recommended by https://specs.openstack.org/openstack/api-wg/guidelines/headers.html

Also:
- add `prettier.config.js` to codify import order
- fix husky path and add husky config (`yarn prepare` failed locally because of cd-ing out of current directory)
- filter out sensitive authorization header from nested request object inside response object in debug logs

### CODE REVIEW

File: client/filter.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import { isArray } from "lodash";
+
+import { createLogger } from "../log";
+
+const logger = createLogger({ name: "filter" });
+
+// jq-node doesn't support `import` syntax
+const jq = require("node-jq");
+
+export const jpFilter = async (
```
Comment (nbrahms) [inline client/filter.ts:10]: I assume you meant
```suggestion
export const jqFilter = async (
```

File: client/filter.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  data: any,
+  jqHeader: string | string[] | undefined
+): Promise<any> => {
+  const query = isArray(jqHeader) ? jqHeader[0] : jqHeader;
```
Comment (nbrahms) [inline client/filter.ts:14]: Wouldn't you want to run this multiple times if there are multiple headers?
```suggestion
  const query = isArray(jqHeader) ? jqHeader : [jqHeader];
```

File: client/filter.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  if (!query) {
+    return data;
+  }
+  try {
+    // jq.run() returns a Promise
```
Comment (nbrahms) [inline client/filter.ts:19]: nit: I think this comment is unnecessary. The fact the line below passes eslint already means the function returns a promise.

File: log/index.ts [hunk 4]
```diff
@@ -22,6 +22,7 @@ export const createLogger = <T extends LoggerOptions>(
     // Redact the authorization header that may contain a secret token
     redact: [
       "response.config.headers.authorization", // Axios intercepted response object
+      "response.request.headers.authorization", // Axios intercepted response object contains the request as well
       "request.headers.authorization", // Axios intercepted request object + forwarded request object (JSON RPC request)
```
Comment (nbrahms) [inline log/index.ts:26]: Do you also want just `headers`? it looks like you manually assign request / response headers to a bare `headers` field in some logs.

File: types/index.ts [hunk 5]
```diff
@@ -2,6 +2,8 @@ import { Request } from "express";
 import core from "express-serve-static-core";
 import { IncomingHttpHeaders } from "node:http";
 
+export const JQ_HEADER = "braekhus-jq-response-filter";
```
Comment (nbrahms) [inline types/index.ts:5]: what motivated you not to add a `x-` prefix? Any specific reason?

Comment (nbrahms) [inline types/index.ts:5]: This is a jq program, not a filter, right? Were it a filter I presume your invocation would look more like `jq.run(". | " + filter, ...)`.

Should this be
```suggestion
export const JQ_HEADER = "x-braekhus-response-jq-program";
```
?

**Review by nbrahms — APPROVED** [review]

File: client/filter.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import { isArray } from "lodash";
+
+import { createLogger } from "../log";
+
+const logger = createLogger({ name: "filter" });
+
+// jq-node doesn't support `import` syntax
```
[the remark this answers is not in the document]
Comment (gergas3) [reply client/filter.ts:7]:
```
Module '"node-jq"' has no exported member 'jq'.
```

File: client/filter.ts (hunk 2 above)
Comment (gergas3) [reply client/filter.ts:14]: You mean chain the header jq commands? The output of the first is the input of the second?
I don't know if that's helpful, I think chains should be expressible in one query.

File: types/index.ts (hunk 5 above)
Comment (gergas3) [reply types/index.ts:5]: It's not recommended anymore:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers

Comment (gergas3) [reply types/index.ts:5]: Called it `braekhus-response-jq-transforrm`

**Pull request merged**